core/src/findReference.ts

```diff
@@ -1,41 +1,42 @@
-import { ConfirmedSignatureInfo, Connection, Finality, PublicKey, SignaturesForAddressOptions } from '@solana/web3.js';
+import { ConfirmedSignatureInfo, Connection, Finality, SignaturesForAddressOptions } from '@solana/web3.js';
+import { Reference } from './types';
 
 /**
  * Thrown when no transaction signature can be found referencing a given public key.
  */
-export class FindTransactionSignatureError extends Error {
-    name = 'FindTransactionSignatureError';
+export class FindReferenceError extends Error {
+    name = 'FindReferenceError';
 }
 
 /**
  * Find the oldest transaction signature referencing a given public key.
  *
- * @throws if signature can't be found
- *
  * @param connection - A connection to the cluster.
- * @param reference - `reference` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#reference)
- * @param {SignaturesForAddressOptions} options - Options for `getSignaturesForAddress`.
- * @param {Finality} finality - A subset of Commitment levels, which are at least optimistically confirmed.
+ * @param reference - `reference` in the [Solana Pay spec](https://github.com/solana-labs/solana-pay/blob/master/SPEC.md#reference).
+ * @param options - Options for `getSignaturesForAddress`.
+ * @param finality - A subset of `Commitment` levels, which are at least optimistically confirmed.
+ *
+ * @throws {FindReferenceError}
  */
-export async function findTransactionSignature(
+export async function findReference(
     connection: Connection,
-    reference: PublicKey,
+    reference: Reference,
     options?: SignaturesForAddressOptions,
     finality?: Finality
 ): Promise<ConfirmedSignatureInfo> {
     const signatures = await connection.getSignaturesForAddress(reference, options, finality);
 
     const length = signatures.length;
-    if (!length) throw new FindTransactionSignatureError('not found');
+    if (!length) throw new FindReferenceError('not found');
 
     // If multiple transaction signatures are found within the limit, return the oldest one.
     if (length < (options?.limit || 1000)) return signatures[length - 1];
 
     try {
         // In the unlikely event that more signatures than the limit are found, recursively find the oldest one.
-        return await findTransactionSignature(connection, reference, options, finality);
+        return await findReference(connection, reference, options, finality);
```

Comment:
Is this an infinite loop since the params are identical? In particular should we be updating `options.before` to the latest signature seen? Basing on https://solana-labs.github.io/solana-web3.js/modules.html#SignaturesForAddressOptions

Reply:
Good catch! I'll fix this.

Reply:
Done.